extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -0,0 +1,326 @@
+      kafkaProperties.setProperty("auto.offset.reset", "smallest");
+
+      consumerConnector = buildConnector(kafkaProperties);
+
+      final List<KafkaStream<String, String>> streams = consumerConnector.createMessageStreamsByFilter(
+          new Whitelist(Pattern.quote(topic)), 1, DEFAULT_STRING_DECODER, DEFAULT_STRING_DECODER
+      );
+
+      if (streams == null || streams.isEmpty()) {
+        throw new IAE("Topic [%s] had no streams", topic);
+      }
+      if (streams.size() > 1) {
+        throw new ISE("Topic [%s] has %d streams! expected 1", topic, streams.size());
+      }
+      final KafkaStream<String, String> kafkaStream = streams.get(0);
+      final ConsumerIterator<String, String> it = kafkaStream.iterator();
+
+      final ListenableFuture<?> future = executorService.submit(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              while (it.hasNext()) {
+                final MessageAndMetadata<String, String> messageAndMetadata = it.next();
+                final String key = messageAndMetadata.key();
+                final String message = messageAndMetadata.message();
+                if (key == null || message == null) {
+                  LOG.error("Bad key/message from topic [%s]: [%s]", topic, messageAndMetadata);
+                  continue;
+                }
+                doubleEventCount.incrementAndGet();
+                map.put(key, message);
+                doubleEventCount.incrementAndGet();
+                LOG.trace("Placed key[%s] val[%s]", key, message);
+              }
+            }
+          }
+      );
+      Futures.addCallback(
+          future, new FutureCallback<Object>()
+          {
+            @Override
+            public void onSuccess(Object result)
+            {
+              LOG.debug("Success listening to [%s]", topic);
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+              if (t instanceof CancellationException) {
+                LOG.debug("Topic [%s] cancelled", topic);
+              } else {
+                LOG.error(t, "Error in listening to [%s]", topic);
+              }
+            }
+          },
+          MoreExecutors.sameThreadExecutor()
+      );
+      this.future = future;
+      started.set(true);
+      return true;
+    }
+  }
+
+  // Overriden in tests
+  ConsumerConnector buildConnector(Properties properties)
+  {
+    return new kafka.javaapi.consumer.ZookeeperConsumerConnector(
+        new ConsumerConfig(properties)
+    );
+  }
+
+  ;
+
+  @Override
+  public boolean close()
+  {
+    synchronized (startStopLock) {
+      if (!started.get() || executorService.isShutdown()) {
+        LOG.info("Already shutdown, ignoring");
+        return false;
+      }
+      started.set(false);
+      executorService.shutdownNow();
+      final ListenableFuture<?> future = this.future;
+      if (future != null) {
```

Comment:
shutdownNow *should* attempt to do a future.cancel(true). This little block is more for ensuring logging of an extreme failure case where the future isn't dying.